### DISCUSSION TITLE
How to compile libcrypto.so and libssl.so with -fvisibility=hidden option ?

### DISCUSSION THREAD

**Discussion by vsrandom** [openssl/openssl#22169] [Q&A] [answered]
In my organization we are right now using openssl 1.1.f. I want to compile libssl.so.1.1 and libcrypto.so.1.1 with the 
-fvisibility=hidden option. 
When I try to add this option with ./config i.e

./config --prefix=$PWD/tmp --openssldir=$PWD/ssl -fvisibility=hidden shared zlib

it leads to linking errors during make command:
...... .libssl.so: undefined reference to `DH_check_params`.

I'm new to compiling stuff in Linux. Will really appreciate some help.

**Comment by levitte** [comment, marked as answer]
Yeah, OpenSSL hasn't caught up with that method for hiding symbols.  We use linker scripts adapted for the different platforms to do the same thing.

Using `-fvisibility=hidden` in that setup produces clashes, 'cause in spite of the linker script for libcrypto, the hidden default means none of the symbols in that library will be exported, so they become unavailable for anything that tries to link with it.  That's why you get the error you get.

Changing the OpenSSL source to use the visibility attributes is a helluva lot of work.  It would certainly be worth looking at at some point, considering all the benefits, and if someone is willing to pick that up, we're willing to consider a PR.

For now,  it's better not to use `-fvisibility=hidden` with OpenSSL's configuration.

**Reply by vsrandom** [reply]
You said - “We use linker scripts adapted for the different platforms to do the same thing.”

Is there a workaround to hide symbols since we can’t use fvisiblity option ?

**Reply by levitte** [reply]
Those linker scripts are that workaround. Simply configure OpenSSL without `-fvisibility=hidden` and you should be fine.

BTW, I forgot to ask, what platform are you doing this on?
